pkg/report/sarif.go

```diff
@@ -0,0 +1,190 @@
+package report
+
+import (
+	"fmt"
+	"html"
+	"io"
+	"regexp"
+	"strings"
+
+	"github.com/owenrumney/go-sarif/v2/sarif"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/fanal/types"
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/vulnerability"
+)
+
+const (
+	sarifOsPackageVulnerability        = "OsPackageVulnerability"
+	sarifLanguageSpecificVulnerability = "LanguageSpecificPackageVulnerability"
+	sarifOtherVulnerability            = "OtherVulnerability"
+	sarifError                         = "error"
+	sarifWarning                       = "warning"
+	sarifNote                          = "note"
+	sarifNone                          = "none"
+)
+
+// regex to extract file path in case string includes (distro:version)
+var re = regexp.MustCompile(`(?P<path>.+?)(?:\s*\((?:.*?)\).*?)?$`)
+
+// SarifWriter implements result Writer
+type SarifWriter struct {
+	Output  io.Writer
+	Version string
+	run     *sarif.Run
+}
+
+type sarifData struct {
+	vulnerabilityId  string
+	fullDescription  string
+	helpText         string
+	helpMarkdown     string
+	resourceType     string
+	severity         string
+	url              string
+	resultIndex      int
+	artifactLocation string
+	message          string
+}
+
+func (sw *SarifWriter) addSarifRule(data *sarifData) {
+	r := sw.run.AddRule(data.vulnerabilityId).
+		WithName(toSarifRuleName(data.resourceType)).
+		WithDescription(data.vulnerabilityId).
+		WithFullDescription(&sarif.MultiformatMessageString{Text: &data.fullDescription}).
+		WithHelp(&sarif.MultiformatMessageString{
+			Text:     &data.helpText,
+			Markdown: &data.helpMarkdown,
+		}).
+		WithDefaultConfiguration(&sarif.ReportingConfiguration{
+			Level: toSarifErrorLevel(data.severity),
+		}).
+		WithProperties(sarif.Properties{
+			"tags": []string{
+				"vulnerability",
+				data.severity,
+			},
+			"precision": "very-high",
+		})
```

Comment:
Should we put severity to `properties.security-severity`?
https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#reportingdescriptor-object

Reply:
sure, I'll add this fiels, but we have to exclude `severity` from `tags`?
I'd leave `tags` unchanged

Reply:
Yes, it would be better to keep it since it can be used for filtering.

Reply:
yes, I've added this field. but i'd like you review for mapping severity to the level of severity (between 0.0 and 10.0).